module/zfs/arc.c

```diff
@@ -5911,32 +5911,20 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 		*arc_flags |= ARC_FLAG_CACHED;
 
 		if (HDR_IO_IN_PROGRESS(hdr)) {
+			zio_t *head_zio = hdr->b_l1hdr.b_acb->acb_zio_head;
 
 			if ((hdr->b_flags & ARC_FLAG_PRIO_ASYNC_READ) &&
 			    priority == ZIO_PRIORITY_SYNC_READ) {
 				/*
-				 * This sync read must wait for an
-				 * in-progress async read (e.g. a predictive
-				 * prefetch).  Async reads are queued
-				 * separately at the vdev_queue layer, so
-				 * this is a form of priority inversion.
-				 * Ideally, we would "inherit" the demand
-				 * i/o's priority by moving the i/o from
-				 * the async queue to the synchronous queue,
-				 * but there is currently no mechanism to do
-				 * so.  Track this so that we can evaluate
-				 * the magnitude of this potential performance
-				 * problem.
-				 *
-				 * Note that if the prefetch i/o is already
-				 * active (has been issued to the device),
-				 * the prefetch improved performance, because
-				 * we issued it sooner than we would have
-				 * without the prefetch.
+				 * This is a sync read that needs to wait for
+				 * an in-flight async read. Request that the
+				 * zio have its priority upgraded.
 				 */
+				ASSERT3P(head_zio, !=, NULL);
+				zio_change_priority(head_zio, priority);
 				DTRACE_PROBE1(arc__sync__wait__for__async,
 				    arc_buf_hdr_t *, hdr);
```

Comment:
@behlendorf I just noticed this dtrace probe. What should I do with it? Should I just rename it (and in `trace_arc.h`)?

Reply:
Yes, let's change the names to match.